### PR TITLE
feature/AKD-138,129: Added Azure 'VM Backups Enabled', 'VM Daily Backup Retention Period' plugin and test cases

### PR DESCRIPTION
**Pull request by ali-imran7**
This PR is similar to #659 so pulled those changes here also to get collector calls and helper changes.

### CODE REVIEW

File: collectors/azure/collector.js [hunk 1]
```diff
@@ -61,6 +61,11 @@ var calls = {
             url: 'https://management.azure.com/subscriptions/{subscriptionId}/providers/Microsoft.KeyVault/vaults?api-version=2019-09-01'
         }
```
Comment (AkhtarAmir) [inline collectors/azure/collector.js:62]:
```suggestion
        },
        listBySubscriptionId: {
            url: 'https://management.azure.com/subscriptions/{subscriptionId}/providers/Microsoft.RecoveryServices/vaults?api-version=2016-06-01'
        }
```

File: collectors/azure/collector.js [hunk 2]
```diff
@@ -61,6 +61,11 @@ var calls = {
...
     },
+    recoveryVaults: {
+        list: {
+            url: 'https://management.azure.com/subscriptions/{subscriptionId}/providers/Microsoft.RecoveryServices/vaults?api-version=2016-06-01'
+        }
+    },
```
Comment (AkhtarAmir) [inline collectors/azure/collector.js:68]:
```suggestion
```

File: helpers/azure/locations.js [hunk 3]
```diff
@@ -53,7 +53,10 @@ module.exports = {
     disks: locations,
     activityLogAlerts: ['global'],
     vaults: locations,
+    recoveryVaults: locations,
```
Comment (AkhtarAmir) [inline helpers/azure/locations.js:56]:
```suggestion
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
+var async = require('async');
+var helpers = require('../../../helpers/azure/');
+
+module.exports = {
+    title: 'VM Daily Backup Retention Period',
+    category: 'Virtual Machines',
+    description: 'Ensures that VM daily backup retention policy is configured to retain backups for desired number of days',
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:7]:
```suggestion
    description: 'Ensure that VM daily backup retention policy is configured to retain backups for the desired number of days.',
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    more_info: 'Azure Backup provides independent and isolated backups to guard against unintended destruction of the data on your VMs. These backups should be retained for a specific amount of time to recover destroyed VM.',
+    recommended_action: 'Configure virtual machine daily backup retention policy to retain backups for desired number of days',
+    link: 'https://docs.microsoft.com/en-us/azure/backup/backup-azure-vms-introduction',
+    apis: ['virtualMachines:listAll', 'recoveryVaults:list', 'backupProtectedItems:list', 'backupPolicies:list'],
+    settings: {
+        daily_backup_retention_period: {
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:13]:
```suggestion
        vm_daily_backup_retention_period: {
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            name: 'VM Daily Backup Retention Period',
+            default: '30',
+            description: 'Number of days that a VM backup will be retained until it is permanently deleted',
+            regex: '^([7-9]|1[0-9]|2[0-9]|30)$'
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:17]:
```suggestion
            description: 'Number of days that a VM backup will be retained until it is permanently deleted',
            regex: '^([7-9]|1[0-9]|2[0-9]|30)$',
            default: '30'
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            const recoveryVaults = helpers.addSource(cache, source,
+                ['recoveryVaults', 'list', location]);
+
+            if (!recoveryVaults || recoveryVaults.err || !recoveryVaults.data) {
+                helpers.addResult(results, 3, 'Unable to query for backup recovery vaults: ' + helpers.addError(recoveryVaults), location);
+                return rcb();
+            }
+
+            if (!recoveryVaults.data.length) {
+                helpers.addResult(results, 2, 'No backup recovery vaults found for the virtual machine', location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:55]:
```suggestion
                helpers.addResult(results, 2, 'No backup recovery vaults found', location);
```

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            const vmPoliciesMap = new Map();
+            const backupPoliciesMap = new Map();
+
+            for (const vault of recoveryVaults.data) {
+                const backupProtectedItems = helpers.addSource(cache, source,
+                    ['backupProtectedItems', 'list', location, vault.id]);
+                const backupPolicies = helpers.addSource(cache, source,
+                    ['backupPolicies', 'list', location, vault.id]);
+
+                if (!backupProtectedItems || backupProtectedItems.err || !backupProtectedItems.data ||
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:68]: Split these checks

File: plugins/azure/virtualmachines/dailyBackupRetention.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+                for (const backupPolicy of backupPolicies.data) {
+                    backupPoliciesMap.set(backupPolicy.id, backupPolicy);
+                }
+            }
+
+            async.each(virtualMachines.data, function(virtualMachine, scb) {
+                const vmPolicies = vmPoliciesMap.get(virtualMachine.id.toLowerCase());
+                if (vmPolicies && vmPolicies.length) {
+                    for (const vmPolicy of vmPoliciesMap.get(virtualMachine.id.toLowerCase())) {
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/dailyBackupRetention.js:91]: Get Policy Map first

File: plugins/azure/virtualmachines/vmBackupsEnabled.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            const recoveryVaults = helpers.addSource(cache, source,
+                ['recoveryVaults', 'list', location]);
+
+            if (!recoveryVaults || recoveryVaults.err || !recoveryVaults.data) {
+                helpers.addResult(results, 3, 'Unable to query for backup vaults: ' + helpers.addError(recoveryVaults), location);
+                return rcb();
+            }
+
+            if (!recoveryVaults.data.length) {
+                helpers.addResult(results, 2, 'No backup vaults found for the virtual machine', location);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmBackupsEnabled.js:43]:
```suggestion
                helpers.addResult(results, 2, 'No backup vaults found', location);
```

File: plugins/azure/virtualmachines/vmBackupsEnabled.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                    for (const policy of vmPolicies) {
+                        if (policy && policy !== '') {
+                            vmBackupsEnabled = true;
+                            break;
+                        }
+                    }
+                }
+
+                if (vmBackupsEnabled) {
+                    helpers.addResult(results, 0, 'Azure virtual machine backups are enabled', location, virtualMachine.id);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmBackupsEnabled.js:82]:
```suggestion
                    helpers.addResult(results, 0, 'Azure virtual machine has backups enabled', location, virtualMachine.id);
```

File: plugins/azure/virtualmachines/vmBackupsEnabled.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+                } else {
+                    helpers.addResult(results, 2, 'Azure virtual machine backups are not enabled', location, virtualMachine.id);
```
Comment (AkhtarAmir) [inline plugins/azure/virtualmachines/vmBackupsEnabled.js:84]:
```suggestion
                    helpers.addResult(results, 2, 'Azure virtual machine does not have backups enabled', location, virtualMachine.id);
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Comment by ali-imran7** [comment]
Need to update collector API calls to make those specific to vaults.

**Pull request closed without merging**